sqlglot/optimizer/qualify_columns.py

```diff
@@ -358,18 +382,26 @@ def _expand_stars(
     replace_columns: t.Dict[int, t.Dict[str, str]] = {}
     coalesced_columns = set()
 
-    # TODO: handle optimization of multiple PIVOTs (and possibly UNPIVOTs) in the future
     pivot_columns = None
     pivot_output_columns = None
-    pivot = t.cast(t.Optional[exp.Pivot], seq_get(scope.pivots, 0))
+    pivot_exclude_columns = None
 
-    has_pivoted_source = pivot and not pivot.args.get("unpivot")
-    if pivot and has_pivoted_source:
-        pivot_columns = set(col.output_name for col in pivot.find_all(exp.Column))
+    pivot = t.cast(t.Optional[exp.Pivot], seq_get(scope.pivots, 0))
+    if isinstance(pivot, exp.Pivot):
+        if pivot.is_unpivot:
+            pivot_output_columns = [c.output_name for c in _unpivot_columns(pivot)]
+
+            field = pivot.args.get("field")
+            if isinstance(field, exp.In):
+                pivot_exclude_columns = {
+                    c.output_name for e in field.expressions for c in e.find_all(exp.Column)
+                }
+        else:
+            pivot_columns = set(c.output_name for c in pivot.find_all(exp.Column))
```

Comment:
why is this a set and 393 a list?

Reply:
this is just speeding up lookups, the lists are used to preserve the order

Reply:
so you do need both variants? they can't both be sets?

Reply:
yep, there's no performance gain in using a set for `pivot_output_columns` whereas iirc using a list has the nice side-effect of _generally_ preserving the order of the columns as returned by the engines when there's a star